ddtrace/internal/ci_visibility/git_client.py

```diff
@@ -0,0 +1,187 @@
+import json
+from multiprocessing import Process
+import os
+import random
+from typing import Optional  # noqa
+from typing import Tuple  # noqa
+
+import tenacity
+
+from ddtrace.ext.git import extract_commit_sha
+from ddtrace.ext.git import extract_latest_commits
+from ddtrace.ext.git import extract_remote_url
+from ddtrace.ext.git import get_rev_list_excluding_commits
+from ddtrace.ext.git import git_subprocess_cmd
+from ddtrace.internal.compat import TemporaryDirectory
+from ddtrace.internal.logger import get_logger
+
+from .. import compat
+from ..utils.http import Response
+from ..utils.http import get_connection
+
+
+log = get_logger(__name__)
+
+
+class CIVisibilityGitClient(object):
+    RETRY_ATTEMPTS = 5
+
+    def __init__(
+        self,
+        api_key,
+        app_key,
+    ):
+        # type: (str, str) -> None
+        self._serde = CIVisibilityGitClientSerDeV1(api_key, app_key)
+        self._worker = None  # type: Optional[Process]
+
+    def start(self, cwd=None):
+        # type: (Optional[str]) -> None
+        if self._worker is None:
+            self._worker = Process(target=CIVisibilityGitClient._run_protocol, args=(self._serde,), kwargs={"cwd": cwd})
+            self._worker.start()
+
+    def shutdown(self, timeout=None):
+        # type: (Optional[float]) -> None
+        if self._worker is not None:
+            self._worker.join(timeout)
+            self._worker = None
+
+    @classmethod
+    def _run_protocol(cls, serde, cwd=None):
+        # type: (CIVisibilityGitClientSerDeV1, Optional[str]) -> None
+        repo_url = cls._get_repository_url(cwd=cwd)
+        latest_commits = cls._get_latest_commits(cwd=cwd)
+        backend_commits = cls._search_commits(repo_url, latest_commits, serde)
+        rev_list = cls._get_filtered_revisions(backend_commits, cwd=cwd)
+        if rev_list:
+            with cls._build_packfiles(rev_list, cwd=cwd) as packfiles_path:
+                cls._upload_packfiles(repo_url, packfiles_path, serde, cwd=cwd)
+
+    @classmethod
+    def _get_repository_url(cls, cwd=None):
+        # type: (Optional[str]) -> str
+        return extract_remote_url(cwd=cwd)
+
+    @classmethod
+    def _get_latest_commits(cls, cwd=None):
+        # type: (Optional[str]) -> list[str]
+        return extract_latest_commits(cwd=cwd)
+
+    @classmethod
+    def _search_commits(cls, repo_url, latest_commits, serde):
+        # type: (str, list[str], CIVisibilityGitClientSerDeV1) -> list[str]
+        payload = serde.search_commits_encode(repo_url, latest_commits)
+        response = cls.retry_request()(cls._do_request, "/search_commits", payload, serde)
+        result = serde.search_commits_decode(response.body)
+        return result
+
+    @classmethod
+    def _do_request(cls, endpoint, payload, serde, headers=None):
+        # type: (str, str, CIVisibilityGitClientSerDeV1, Optional[dict]) -> Response
+        url = "https://git-api-ci-app-backend.us1.staging.dog/repository%s" % endpoint
```

Comment:
is this URL correct?

Reply:
@juan-fernandez No, it's a staging URL. Do you happen to know the correct production URL?

Reply:
the base URL is the same as the one for ITR config and skippable calls: `https://api.${DD_SITE}`